### ISSUE TITLE
resque + devise = no emails

### ISSUE THREAD

**Issue by jonbcampos** [plataformatec/devise#4098]
I'm reading through the readme and there must be something I'm missing.

[https://github.com/plataformatec/devise#activejob-integration](https://github.com/plataformatec/devise#activejob-integration)

I have three different user models (shouldn't be a problem), and I did get emails at one point through the signup process.

I've since added resque and resque_mailer. If I send emails that are not from devise the emails go out no problem via resque_mailer. Any devise emails show to have been added to the queue but stop there. I don't see any email errors when I add "raise email errors" in my environment variables.

The following is the line output from papertrail with the activejob details:

[ActiveJob] [ActionMailer::DeliveryJob] [0e404d2f-0ae9-4365-8718-8a0d3358eb60] Performing ActionMailer::DeliveryJob from Resque(mailers) with arguments: "Devise::Mailer", "confirmation_instructions", "deliver_now", gid://app/User/2, "ECVvcJxGPyv5wDRFpw4W", {} 

any ideas?

**Comment by jonbcampos** [comment]
fyi, it works on test, it works on develop (writing file), it just doesn't work with sendgrid on heroku

**Comment by jonbcampos** [comment]
@ulissesalmeida what kind of info would be helpful. I don't want to dump a ton of info that is worthless.

I am using resque and resque_mailer. Resque mailer has you create a base class mailer (mine is called 'AsyncMailer') that includes Resque::Mailer. I've tried putting this as the Devise parent_mailer

    Devise.setup do |config|
        config.parent_mailer = 'AsyncMailer'
    end

When I debug on test or development I can see the mailer being created and looking correct (but doesn't actually send):

    protected
    def send_devise_notification(notification, *args)
        mail = devise_mailer.send(notification, self, *args)
        mail.deliver_in(1.minute)
     end

Produces a similar mail object as a working segment of code:

     mailer = AdminMailer.on_the_fly(@to, from, @subject, @message)
     mailer.deliver_in(@minutes_from_now.minutes)

Thoughts? more info needed?

**Comment by jonbcampos** [comment]
I've also tried and failed with:

    protected
    def send_devise_notification(notification, *args)
        mail = devise_mailer.send(notification, self, *args)
        mail.deliver
     end

**Comment by jonbcampos** [comment]
From the logs it appears to send... and then doesn't:

May 10 14:34:11 myapp app/web.1:  Started POST "/users/password" for 198.179.137.229 at 2016-05-10 21:34:10 +0000 
May 10 14:34:11 myapp app/web.1:  Processing by Users::PasswordsController#create as HTML 
May 10 14:34:11 myapp app/web.1:    Parameters: {"utf8"=>"✓", "authenticity_token"=>"[my token]", "user"=>{"email"=>"user@myapp.com"}, "commit"=>"Send Me Reset Password Instructions"} 
May 10 14:34:12 myapp heroku/router:  at=info method=GET path="/users/sign_in" host=myapp.herokuapp.com request_id=93e1f19e-1ba9-49e8-a77a-e2ab3a320dc0 fwd="198.179.137.229" dyno=web.1 connect=1ms service=131ms status=200 bytes=9484 
May 10 14:34:12 myapp heroku/router:  at=info method=POST path="/users/password" host=myapp.herokuapp.com request_id=951509d5-ac19-4910-bb97-cc07a4341a71 fwd="198.179.137.229" dyno=web.1 connect=1ms service=1421ms status=302 bytes=710 
May 10 14:34:12 myapp app/web.1:  Redirected to https://myapp.herokuapp.com/users/sign_in 
May 10 14:34:12 myapp app/web.1:  Completed 302 Found in 1394ms (ActiveRecord: 35.5ms) 
May 10 14:34:12 myapp app/web.1:  Started GET "/users/sign_in" for 198.179.137.229 at 2016-05-10 21:34:11 +0000 
May 10 14:34:12 myapp app/web.1:  Processing by Users::SessionsController#new as HTML 
May 10 14:34:12 myapp app/web.1:  Completed 200 OK in 78ms (Views: 71.0ms | ActiveRecord: 0.0ms) 
May 10 14:34:14 myapp app/resque.1:  *** got: (Job{mailer} | Devise::Mailer | ["reset_password_instructions", [{"class_name"=>"User", "id"=>1}, "aExTZsJBPiTNR5qs3s9a", {}]]) 
May 10 14:34:14 myapp app/resque.1:  *** Running before_fork hooks with [(Job{mailer} | Devise::Mailer | ["reset_password_instructions", [{"class_name"=>"User", "id"=>1}, "aExTZsJBPiTNR5qs3s9a", {}]])] 
May 10 14:34:14 myapp app/resque.1:  *** Running after_fork hooks with [(Job{mailer} | Devise::Mailer | ["reset_password_instructions", [{"class_name"=>"User", "id"=>1}, "aExTZsJBPiTNR5qs3s9a", {}]])] 
May 10 14:34:16 myapp heroku/router:  at=info method=GET path="/login" host=myapp.herokuapp.com request_id=ba893bd4-8451-477f-a2c5-eb754ed5a137 fwd="198.179.137.229" dyno=web.1 connect=1ms service=53ms status=200 bytes=7352 
May 10 14:34:16 myapp app/web.1:  Started GET "/login" for 198.179.137.229 at 2016-05-10 21:34:15 +0000 
May 10 14:34:16 myapp app/web.1:  Processing by PagesController#login as HTML 
May 10 14:34:16 myapp app/web.1:  Completed 200 OK in 30ms (Views: 27.1ms | ActiveRecord: 0.0ms) 
May 10 14:34:16 myapp heroku/router:  at=info method=GET path="/admins/sign_in" host=myapp.herokuapp.com request_id=24ae3229-0303-483b-b9ae-c9a4596a0e96 fwd="198.179.137.229" dyno=web.1 connect=1ms service=63ms status=200 bytes=9077 
May 10 14:34:16 myapp app/resque.1:  Sent mail to user@myapp.com (272.0ms) 
May 10 14:34:16 myapp app/resque.1:  *** done: (Job{mailer} | Devise::Mailer | ["reset_password_instructions", [{"class_name"=>"User", "id"=>1}, "aExTZsJBPiTNR5qs3s9a", {}]]) 
May 10 14:34:17 myapp app/web.1:  Started GET "/admins/sign_in" for 198.179.137.229 at 2016-05-10 21:34:16 +0000

**Comment by ulissesalmeida** [comment]
@jonbcampos I'm sorry I've put the label, but I did forget to comment. My bad.

Usually, when we put `Needs more info` is because we need the steps to reproduce the error to check if it is a Devise issue.

Since you're trying to use the Devise with Resque is hard to figure out where the problem is, it could be on `Devise`, `Resque`, the integration between both, or maybe the e-mail sender service.

Looking at the logs, 🔍 , for example:

```
May 10 14:34:16 myapp app/web.1:  Completed 200 OK in 30ms (Views: 27.1ms | ActiveRecord: 0.0ms) 
May 10 14:34:16 myapp heroku/router:  at=info method=GET path="/admins/sign_in" host=myapp.herokuapp.com request_id=24ae3229-0303-483b-b9ae-c9a4596a0e96 fwd="198.179.137.229" dyno=web.1 connect=1ms service=63ms status=200 bytes=9077 
May 10 14:34:16 myapp app/resque.1:  Sent mail to user@myapp.com (272.0ms) 
May 10 14:34:16 myapp app/resque.1:  *** done: (Job{mailer} | Devise::Mailer | ["reset_password_instructions", [{"class_name"=>"User", "id"=>1}, "aExTZsJBPiTNR5qs3s9a", {}]]) 
```

It looks no error happened. It seems ok. If you can isolate and reproduce the problem in a sample application and provide the GitHub link to us, it will be awesome. We can debug the app try to figure out what is happening.

**Comment by jonbcampos** [comment]
So... good news! I'm closing this. After WAY WAY WAY too much investigation I figured out that the issue was with my hoster and DNS.

It wasn't obvious at first till I found my sendgrid activity logs and saw the 550 errors. Then I traced the error, switched DNS providers (something I needed to do at some point anyways) and fixed the issue.

Devise and Resque work exactly as described in the documentation.

Check your logs, debugging this issue took way too long.
